src/VecSim/algorithms/hnsw/hnswlib.h

```diff
@@ -1269,4 +1341,85 @@ HierarchicalNSW<dist_t>::searchKnn(const void *query_data, size_t k, void *timeo
     return results;
 }
 
+template <typename dist_t>
+VecSimQueryResult *HierarchicalNSW<dist_t>::searchRangeBottomLayer_WithTimeout(
+    tableint ep_id, const void *data_point, float epsilon, float radius, void *timeoutCtx,
+    VecSimQueryResult_Code *rc) const {
+    auto *results = array_new<VecSimQueryResult>(10); // arbitrary initial cap.
+
+#ifdef ENABLE_PARALLELIZATION
+    this->visited_nodes_handler =
+        this->visited_nodes_handler_pool->getAvailableVisitedNodesHandler();
+#endif
+
+    tag_t visited_tag = this->visited_nodes_handler->getFreshTag();
+    candidatesMaxHeap<dist_t> candidate_set(this->allocator);
+
+    // Set the initial effective-range to be at least the distance from the entry-point.
+    dist_t ep_dist = fstdistfunc_(data_point, getDataByInternalId(ep_id), dist_func_param_);
+    dist_t dynamic_range = ep_dist;
+    if (ep_dist <= radius) {
+        // Entry-point is within the radius - add it to the results.
+        auto new_result = VecSimQueryResult{};
+        VecSimQueryResult_SetId(new_result, getExternalLabel(ep_id));
+        VecSimQueryResult_SetScore(new_result, ep_dist);
+        results = array_append(results, new_result);
+        dynamic_range = radius; // to ensure that dyn_range >= radius.
+    }
+
+    candidate_set.emplace(-ep_dist, ep_id);
+    this->visited_nodes_handler->tagNode(ep_id, visited_tag);
+
+    while (!candidate_set.empty()) {
+        std::pair<dist_t, tableint> curr_el_pair = candidate_set.top();
+        // If the best candidate is outside the dynamic range in more than epsilon (relatively) - we
+        // finish the search.
+        if ((-curr_el_pair.first) > dynamic_range * (1.0f + epsilon)) {
+            break;
+        }
+        if (__builtin_expect(VecSimIndex::timeoutCallback(timeoutCtx), 0)) {
+            *rc = VecSim_QueryResult_TimedOut;
+            return results;
+        }
+        candidate_set.pop();
+
+        // Decrease the effective range, but keep dyn_range >= radius.
+        if (-curr_el_pair.first < dynamic_range && -curr_el_pair.first > radius) {
+            dynamic_range = -curr_el_pair.first;
+        }
```

Comment:
consider having `dynamic_range` and `search_range` or other name for the range and the calculated one with the epsilon.